primitives/src/vstaging/mod.rs

```diff
@@ -50,3 +53,242 @@ pub struct AsyncBackingParams {
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, RuntimeDebug)]
+pub struct CollatorRestrictions {
+	/// Collators to prefer/allow.
+	/// Empty set means no restrictions.
+	collator_peer_ids: BTreeSet<OpaquePeerId>,
+	restriction_kind: CollatorRestrictionKind,
+}
+
+impl CollatorRestrictions {
+	/// Specialised new function for parachains.
+	pub fn none() -> Self {
+		CollatorRestrictions {
+			collator_peer_ids: BTreeSet::new(),
+			restriction_kind: CollatorRestrictionKind::Preferred,
+		}
+	}
+
+	/// Create a new `CollatorRestrictions`.
+	pub fn new(
+		collator_peer_ids: BTreeSet<OpaquePeerId>,
+		restriction_kind: CollatorRestrictionKind,
+	) -> Self {
+		CollatorRestrictions { collator_peer_ids, restriction_kind }
+	}
+
+	/// Is peer_id allowed to collate?
+	#[cfg(feature = "std")]
+	pub fn can_collate(&self, peer_id: &PeerId) -> bool {
+		self.collator_peer_ids.is_empty() ||
+			match self.restriction_kind {
+				CollatorRestrictionKind::Preferred => true,
+				CollatorRestrictionKind::Required => {
+					let peer_id = OpaquePeerId(peer_id.to_bytes());
+					self.collator_peer_ids.contains(&peer_id)
+				},
+			}
+	}
+}
+
+/// How to apply the collator restrictions.
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, RuntimeDebug)]
+pub enum CollatorRestrictionKind {
+	/// peer ids mentioned will be preferred in connections, but others are still allowed.
+	Preferred,
+	/// Any collator with a `PeerId` not in the set of `CollatorRestrictions` will be rejected.
+	Required,
+}
+
+/// An entry tracking a paras
+#[derive(Clone, Encode, Decode, TypeInfo, PartialEq, RuntimeDebug)]
+pub struct ParasEntry {
+	/// The `Assignment`
+	pub assignment: Assignment,
+	/// Number of times this has been retried.
+	pub retries: u32,
+}
+
+impl From<Assignment> for ParasEntry {
+	fn from(assignment: Assignment) -> Self {
+		Self::new(assignment)
+	}
+}
+
+impl ParasEntry {
+	/// Create a new `ParasEntry`.
+	pub fn new(assignment: Assignment) -> Self {
+		ParasEntry { assignment, retries: 0 }
+	}
+
+	/// Return `Id` from the underlying `Assignment`.
+	pub fn para_id(&self) -> Id {
+		self.assignment.para_id
+	}
+
+	/// Return `CollatorRestrictions` from the underlying `Assignment`.
+	pub fn collator_restrictions(&self) -> &CollatorRestrictions {
+		&self.assignment.collator_restrictions
+	}
+}
+
+/// Information about a core which is currently occupied.
+#[derive(Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct OccupiedCore<H = Hash, N = BlockNumber> {
+	// NOTE: this has no ParaId as it can be deduced from the candidate descriptor.
+	/// If this core is freed by availability, this is the assignment that is next up on this
+	/// core, if any. None if there is nothing queued for this core.
+	pub next_up_on_available: Option<ScheduledCore>,
+	/// The relay-chain block number this began occupying the core at.
+	pub occupied_since: N,
+	/// The relay-chain block this will time-out at, if any.
```

Comment:
This is not an `Option`, so the `if any` seems odd.

Reply:
This is a copy and paste from `v4`. Fixed.